### ISSUE TITLE
Very slow bundling performance with sourcemaps

### ISSUE THREAD

**Issue by jackfranklin** [systemjs/builder#253]
As you know, at GC we're now building every time a file is changed and due to the caching we can get this going in about 2 seconds max, however without source maps.

I've been trying to add sourcemaps but the moment we turn them on, even with the low res option, the time jumps through the roof. Some very basic numbers below:

```
{ sourceMaps: false, minify: false }
- From fresh: ~6 secs
- From cache: ~2 secs

{ sourceMaps: true, minify: false }
- From fresh: ~40 seconds
- From cache: ~20 seconds

{ sourceMaps: true, lowResSourceMaps: true, minify: false }
- From fresh: ~17 seconds
- From cache: ~4 seconds

{ sourceMaps: true, lowResSouceMaps: true, minify: true }
- From fresh: ~45 seconds
- From cache: ~20 seconds
```

It would be awesome to dig into why this is and if there's any similar caching type approach that might save some time here. I presume the delay is because the sitemap has to be regenerated from scratch after the bundle is built?

**Comment by guybedford** [comment]
The source maps are actually being cached here, it's just the process of combining them all that takes time.

Of the 3 source maps cases above, which one in particular are you looking to optimize? It may be advisable to go with the `lowResSourceMaps` option?

**Comment by guybedford** [comment]
Actually I assume you've got minification off for development, so it really is just the source maps combination code you're looking to optimize here?

**Comment by jackfranklin** [comment]
@guybedford we don't minify in development, no, but in an ideal world we'd have a sourcemap still pointing back to the original source. If there's not much that can be done to speed this up then that's not the end of the world, I just more wanted to know if there was something we could do.

>so it really is just the source maps combination code you're looking to optimize here

Basically, yes! :dancer:

**Comment by guybedford** [comment]
There's always optimizations that can be done - I guess we could make sure that the cached source map representation is a serialized `SourceMapConsumer` class that can be quickly reconstructed from the cached serialization, which will save the performance costs of reparsing all source maps each time. I've created https://github.com/systemjs/builder/issues/254 to track this.

The final output generation has already been significantly optimized by @crisptrutski where we implemented the `lowResSourceMaps` as an option as well here.

For minifiation, we can improve the time via per-output minification caching tracking in https://github.com/systemjs/builder/issues/209.

Those are the main things I can think of for now on this issue.
